### PR TITLE
Added sub for bignum

### PR DESCRIPTION
**Pull request by luke-bhan**
Added sub method. I didn't end up reversing the dec vector since it would have ruined our ability to use truncate in removing ending zeros from the decmial. Added utils in bignum for the is_smaller method, we may need this in other files like divide in the future so I moved it now. lmk what you think @ayazhafiz

### CODE REVIEW

File: libslide/src/bignum.rs [hunk 1]
```diff
@@ -24,6 +26,11 @@ impl Bignum {
         let mut has_decimal = false;
         let mut int = Vec::new();
         let mut dec = Vec::new();
+
+        if input.is_empty() {
+            panic!("A arbitray precision number must have a valid input");
+        }
+
```
Comment (ayazhafiz) [inline libslide/src/bignum.rs:33]: Maybe it's better to return an option, but we can do that in a different pr

File: libslide/src/bignum.rs [hunk 2]
```diff
@@ -60,8 +67,12 @@ impl fmt::Display for Bignum {
         if self.is_neg {
             result.push('-');
         }
-        for n in self.int.iter().rev() {
-            result.push(to_char(*n));
+        if self.int.is_empty() {
+            result.push('0');
```
Comment (ayazhafiz) [inline libslide/src/bignum.rs:71]: Why?

File: libslide/src/bignum/add.rs [hunk 3]
```diff
@@ -23,6 +23,7 @@ impl ops::Add for Bignum {
             (rhs.dec, self.dec)
         };
 
+        // 1. Handle Decimal
```
Comment (ayazhafiz) [inline libslide/src/bignum/add.rs:26]: :+1:

File: libslide/src/bignum/sub.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            if res_int[i] == 0 {
+                count += 1;
+            } else {
+                break;
+            }
+        }
+
+        res_int.truncate(res_int.len() - count);
+
+        print!("{:?}", res_dec);
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:92]: dd

File: libslide/src/bignum/sub.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    type Output = Bignum;
+
+    fn sub(self, other: Bignum) -> Bignum {
+        // make the decimal vectors the same size
+        // fill lhs and rhs vectors as well
+        let mut lhs_dec: Vec<u8>;
+        let mut res_dec: Vec<u8>;
+        let self_size: usize = self.dec.len();
+        let other_size: usize = other.dec.len();
+        let (mut res_int, mut lhs_int) = if self_size > other_size {
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:18]: We can set all four lhs_, res_ vecs here!

File: libslide/src/bignum/sub.rs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            res_dec = self.dec;
+            lhs_dec = vec![0; self_size];
+            lhs_dec[..other_size].clone_from_slice(&other.dec[..other_size]);
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:21]: I'm confused by this. Why not just
```suggestion
            lhs_dec = other.dec
```

File: libslide/src/bignum/sub.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+use std::ops;
+
+impl ops::Sub for Bignum {
+    type Output = Bignum;
+
+    fn sub(self, other: Bignum) -> Bignum {
+        // make the decimal vectors the same size
+        // fill lhs and rhs vectors as well
+        let mut lhs_dec: Vec<u8>;
+        let mut res_dec: Vec<u8>;
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:15]: I think it would be cheaper to set up the vecs as `i8`s here to avoid casting in a loop. LLVM is probably smart enough to optimize it out anyway, but here we make it one step closer and the optimizer can probably vectorize the casts.

File: libslide/src/bignum/sub.rs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            lhs_dec = vec![0; other_size];
+            lhs_dec[..self_size].clone_from_slice(&self.dec[..self_size]);
+            (other.int, self.int)
+        };
+
+        // now we need to make the bigger number the res dec
+        if is_smaller(&lhs_dec, &res_dec, &lhs_int, &res_int) {
+            mem::swap(&mut lhs_dec, &mut res_dec);
+            mem::swap(&mut lhs_int, &mut res_int);
+        }
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:34]: Why not implement [Ord](https://doc.rust-lang.org/std/cmp/trait.Ord.html) for bignum instead?

File: libslide/src/bignum/sub.rs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+        // 1. Decimal
+
+        let mut res: i8;
+        let mut carry: i8 = 0;
+        for i in (0..res_dec.len()).rev() {
+            res = res_dec[i] as i8 - lhs_dec[i] as i8 - carry;
+            // not sure if you can use a let if here because of shadowing
+            // but maybe you know
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:43]: you can't use a let if, but note that an if can be an expression. That is, you can do something like `a = if true { 1 } else { 2 }`

File: libslide/src/bignum/sub.rs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+            if res < 0 {
+                res += 10;
+                carry = 1;
+            } else {
+                carry = 0;
+            }
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:49]: This is not the prettiest thing in the world, but it does save a branching:

```suggestion
            carry = (res < 0) as i8;
            res += carry * 10; // add 10 if negative, nothing otherwise
```

File: libslide/src/bignum/sub.rs (hunk 6 above)
Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs:21]: Got it, nevermind. Is it possible to move a slice into another? If not another option is to set `lhs_dec = other.dec`, then `.reserve()` the rest of the space needed and add 0s. This is pedantic though, the cost savings are just copying some 8-bit values and extra zeros.

File: libslide/src/bignum/sub.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+        for i in 0..lhs_int.len() {
+            res = res_int[i] as i8 - lhs_int[i] as i8 - carry;
+            if res < 0 {
+                res += 10;
+                carry = 1;
+            } else {
+                carry = 0;
+            }
+            res_int[i] = res.try_into().unwrap();
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:63]: Yeah let's make res_int a `Vec<i8>`

File: libslide/src/bignum.rs (hunk 2 above)
Comment (luke-bhan) [reply libslide/src/bignum.rs:71]: I want decimals to be printed 0.1 instead of .1

File: libslide/src/bignum/sub.rs (hunk 8 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:34]: I will do this instead :)

File: libslide/src/bignum/sub.rs (hunk 9 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:43]: oh sick

File: libslide/src/bignum/sub.rs (hunk 11 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:63]: So technically this wont change the casting. Because either way res_int has to be a u8 in Bignum. so we would have to convert to build it and then convert back to store it

File: libslide/src/bignum.rs (hunk 1 above)
Comment (luke-bhan) [reply libslide/src/bignum.rs:33]: how would we retunr an option in the constructor

Comment (ayazhafiz) [reply libslide/src/bignum.rs:33]: Just change the return type of the method

File: libslide/src/bignum.rs (hunk 2 above)
Comment (ayazhafiz) [reply libslide/src/bignum.rs:71]: +1

File: libslide/src/bignum/sub.rs (hunk 11 above)
Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs:63]: that's fine, it's cleaner than doing it each time in a loop imo

Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:63]: so how do you wanna build res_dec. fill with zeros and then cast? I think its equally ugly

Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:63]: also converting to a vector creates a  new allocation of a vector. I think I like this way better

Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs:63]: just do it this way: https://stackoverflow.com/questions/59707349/cast-vector-of-i8-to-vector-of-u8-in-rust

**Comment by luke-bhan** [comment]
@ayazhafiz I updated the PR. LMK what you think.

File: libslide/src/bignum/compare.rs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        if self.dec.len() != other.dec.len() {
+            return false;
+        }
+        for i in 0..other.dec.len() {
+            if other.dec[i] != self.dec[i] {
+                return false;
+            }
+        }
+        true
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:62]: Why not just

```rust
self.cmp(other) == Ordering::Equal
```

File: libslide/src/bignum/sub.rs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        let mut lhs = Bignum {
+            is_neg: false,
+            dec: lhs_dec,
+            int: lhs_int,
+        };
+        let mut rhs = Bignum {
+            is_neg: false,
+            dec: res_dec,
+            int: res_int,
+        };
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:37]: let's not construct a new bignum just to do the swap

File: libslide/src/bignum/utils.rs [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    unsafe { Vec::from_raw_parts(p as *mut u8, len, cap) }
+}
+
+pub fn vec_u8_into_i8(v: Vec<u8>) -> Vec<i8> {
+    let mut v = std::mem::ManuallyDrop::new(v);
+    let p = v.as_mut_ptr();
+    let len = v.len();
+    let cap = v.capacity();
+    unsafe { Vec::from_raw_parts(p as *mut i8, len, cap) }
+}
```
Comment (ayazhafiz) [inline libslide/src/bignum/utils.rs:15]: We can combine this

```suggestion
pub fn recast_vec<T, U>(v: Vec<T>) -> Vec<U> {
    let mut v = std::mem::ManuallyDrop::new(v);
    let p = v.as_mut_ptr();
    let len = v.len();
    let cap = v.capacity();
    unsafe { Vec::from_raw_parts(p as *mut U, len, cap) }
}
```

File: libslide/src/bignum/sub.rs [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        // Propogate the rest of the carry
+        let mut i = lhs_int_asi8.len();
+        while i < res_int_asi8.len() {
+            res = res_int_asi8[i] as i8 - carry;
+            if res < 0 {
+                res += 10;
+                carry = 1;
+            } else {
+                carry = 0;
+            }
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:74]: Suggestion from before

```suggestion
carry = res < 0;
res += 10 * carry;
```

File: libslide/src/bignum/sub.rs [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            if res_dec_asi8[i] == 0 {
+                count += 1;
+            } else {
+                break;
+            }
+        }
+        res_dec_asi8.truncate(res_dec_asi8.len() - count);
+
+        Bignum {
+            is_neg: false,
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:102]: Do you plan to handle negatives in another PR?

**Comment by ayazhafiz** [comment]
@luke-bhan left some comments, mostly minor stuff

File: libslide/src/bignum/sub.rs (hunk 16 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:102]: Yes

File: libslide/src/bignum/sub.rs (hunk 13 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:37]: It’s not to do the swap, it’s to compare. What’s a better way of doing this then.

Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs:37]: nevermind, sounds good 👍

Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs:37]: Actually, hold up. Both self and rhs are moved in at the beginning of this function, so why not just compare them then before you construct the vectors?

File: libslide/src/bignum/compare.rs (hunk 12 above)
Comment (luke-bhan) [reply libslide/src/bignum/compare.rs:62]: What do you mean

File: libslide/src/bignum/sub.rs (hunk 13 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:37]: That’s how I originally wrote sub, but Then you run into an issue with adding the zeros to make the decimals the same length.

Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs:37]: What's wrong with that?

File: libslide/src/bignum/compare.rs (hunk 12 above)
Comment (ayazhafiz) [reply libslide/src/bignum/compare.rs:62]: Replace this whole method with `self.cmp(other) == Ordering::Equal`

**Comment by luke-bhan** [comment]
ok, @ayazhafiz everything should be fixed except tests for ordering. I will add that and negatives in a quick fixup pr in a few hours (I don't know how long its gonna take hopefully less)

**Comment by luke-bhan** [comment]
Ok, I actually added negatives and compares in this pr. (I'm lazy and don't want to make an extra one lol)

File: libslide/src/bignum/add.rs [hunk 17]
```diff
@@ -7,7 +8,14 @@ impl ops::Add for Bignum {
     type Output = Bignum;
 
     fn add(self, rhs: Bignum) -> Bignum {
-        let is_neg: bool = rhs.is_neg && self.is_neg;
+        let mut is_neg: bool = false;
+        if self.is_neg && rhs.is_neg {
+            is_neg = true;
+        } else if self.is_neg && !rhs.is_neg {
+            return rhs - abs(&self);
+        } else if !self.is_neg && rhs.is_neg {
```
Comment (ayazhafiz) [inline libslide/src/bignum/add.rs:16]: You only need to do one comparison on each branch
```suggestion
        } else if self.is_neg {
            return rhs - abs(&self);
        } else if rhs.is_neg {
```

File: libslide/src/bignum/add.rs [hunk 18]
```diff
@@ -7,7 +8,14 @@ impl ops::Add for Bignum {
     type Output = Bignum;
 
     fn add(self, rhs: Bignum) -> Bignum {
-        let is_neg: bool = rhs.is_neg && self.is_neg;
+        let mut is_neg: bool = false;
+        if self.is_neg && rhs.is_neg {
+            is_neg = true;
```
Comment (ayazhafiz) [inline libslide/src/bignum/add.rs:13]:
```suggestion
        let mut is_neg: bool = self.is_neg && rhs.is_neg;
        if self.is_neg && rhs.is_neg {
            is_neg = true;
```

And just get rid of this branch

File: libslide/src/bignum/compare.rs [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            return Ordering::Less;
+        } else if other.is_neg && !self.is_neg {
+            return Ordering::Greater;
+        } else if other.is_neg && self.is_neg {
+            greater = Ordering::Less;
+            less = Ordering::Greater;
+        } else {
+            greater = Ordering::Greater;
+            less = Ordering::Less;
+        }
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:18]: Why include these branches?

File: libslide/src/bignum/compare.rs [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+use std::cmp::Ordering;
+
+impl Ord for Bignum {
+    fn cmp(&self, other: &Bignum) -> Ordering {
+        let greater: Ordering;
+        let less: Ordering;
+        if self.is_neg && !other.is_neg {
+            return Ordering::Less;
+        } else if other.is_neg && !self.is_neg {
+            return Ordering::Greater;
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:11]: Do you want to see a way to save a branch in a somewhat convoluted manner? Great, you're in luck!
```suggestion
        if self.is_neg ^ other.is_neg {
             return other.is_neg.cmp(&self.is_neg)
        }
```

File: libslide/src/bignum/compare.rs [hunk 21]
```diff
@@ -0,0 +1,131 @@
+use crate::bignum::Bignum;
+use std::cmp::Ordering;
+
+impl Ord for Bignum {
+    fn cmp(&self, other: &Bignum) -> Ordering {
+        let greater: Ordering;
+        let less: Ordering;
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:7]: nit: prefer using `Ordering::` directly rather than variables

File: libslide/src/bignum/compare.rs [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+                    _ => panic!("Test input invalid"),
+                }
+            }
+        )*
+        }
+    }
+
+    mod cmp {
+        bignum_test_cmp! {
+            equal1: "1", "1", "="
```
Comment (ayazhafiz) [inline libslide/src/bignum/compare.rs:96]: Beautiful! Just one suggestion - write the test as `1 = 1` and split by spaces into three components to compare with.

File: libslide/src/bignum/sub.rs [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+
+        // now we need to make the bigger number the res dec
+        // 1. Decimal
+
+        let mut res: i8;
+        let mut carry: i8 = 0;
+        let mut res_dec_asi8 = recast_vec::<u8, i8>(res_dec);
+        let lhs_dec_asi8 = recast_vec::<u8, i8>(lhs_dec);
+        let mut res_int_asi8 = recast_vec::<u8, i8>(res_int);
+        let lhs_int_asi8 = recast_vec::<u8, i8>(lhs_int);
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:57]: I think it would make more sense to do this in lines 39-46.

File: libslide/src/bignum/sub.rs [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+            if self.is_neg {
+                is_neg = true;
+            }
+            lhs = other;
+            rhs = self;
+        } else {
+            // this really is !self.is_neg && !other.is_neg but we can use demorgans for speed
+            if !(self.is_neg || other.is_neg) {
+                is_neg = true;
+            }
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:30]: It would be even faster if you set this on line 11 and avoided the branch.

File: libslide/src/bignum/sub.rs [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+
+        // quick optimization to remove -0 from possible results.
+        if is_neg && res_int_asi8.is_empty() && res_dec_asi8.is_empty() {
+            is_neg = false;
+        }
+
+        Bignum {
+            is_neg,
+            dec: recast_vec::<i8, u8>(res_dec_asi8),
+            int: recast_vec::<i8, u8>(res_int_asi8),
```
Comment (ayazhafiz) [inline libslide/src/bignum/sub.rs:114]: I think you can elide the type parameters

File: libslide/src/bignum/compare.rs (hunk 19 above)
Comment (luke-bhan) [reply libslide/src/bignum/compare.rs:18]: because if its a negative number we need to switch the comparison. because we are comparing the magnitude in all the branches? I know this isn't a great way to handle it but I couldn't really think of anything cleaner. Do you have any general ideas?

File: libslide/src/bignum/sub.rs (hunk 25 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:114]: how? doesn't it need to know what type to cast to?

File: libslide/src/bignum/sub.rs (hunk 23 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:57]: yep. will do

File: libslide/src/bignum/sub.rs (hunk 24 above)
Comment (luke-bhan) [reply libslide/src/bignum/sub.rs:30]: you can't do this, I am already avoiding branches by setting it to false on line 11

File: libslide/src/bignum/compare.rs (hunk 19 above)
Comment (ayazhafiz) [reply libslide/src/bignum/compare.rs:16]: nvm this is fine

File: libslide/src/bignum/sub.rs (hunk 25 above)
Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs]: Try it

File: libslide/src/bignum/sub.rs (hunk 24 above)
Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs]: Why can't you do it? Just do is_neg = self.is_neg || other.is_neg.

Comment (luke-bhan) [reply libslide/src/bignum/sub.rs]: oh i see ok.

Comment (luke-bhan) [reply libslide/src/bignum/sub.rs]: I was thinking  a different idea

Comment (luke-bhan) [reply libslide/src/bignum/sub.rs]: this idea doesn't work. It needs to be a specific case where the lhs is < rhs to work. I think the way I have it is fine.

Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs]: How can that be the case? If they're both negative numbers the result must be negative.

Comment (luke-bhan) [reply libslide/src/bignum/sub.rs]: no, this is handling if  they are both positive and lhs is < rhs.

Comment (ayazhafiz) [reply libslide/src/bignum/sub.rs]: got it
